### ISSUE TITLE
ICE: 'assertion failed: `(left != right)` at compiler/rustc_parse/src/parser/ty.rs:337:9

### ISSUE THREAD

**Issue by Cancelll** [rust-lang/rust#86083]
### Code

```Rust
# ! [ feature ( nll ) ]
fn doit ( data : & ( for < > r#𐀥?혜ଡ଼เև𐀹 ( ) ) ) {
|| doit ( data ) ;
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.54.0-nightly (dbe459ded 2021-06-02)
binary: rustc
commit-hash: dbe459dedd33470f2cb28101157de316caaffa66
commit-date: 2021-06-02
host: x86_64-unknown-linux-gnu
release: 1.54.0-nightly
LLVM version: 12.0.1
```

### Error output

```
thread 'rustc' panicked at 'assertion failed: `(left != right)`
  left: `Token { kind: Question, span: default_finding_folder_rust/crash_20210605_162043_0093/delta/perses_result/mutant.rs:2:33: 2:34 (#0) }`,
 right: `Question`', compiler/rustc_parse/src/parser/ty.rs:337:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.54.0-nightly (dbe459ded 2021-06-02) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z mir-opt-level=3 -Z new-llvm-pass-manager=yes -Z unsound-mir-opts=yes -Z polonius=yes -Z polymorphize=yes -C debuginfo=2 -C opt-level=3 --crate-type staticlib

query stack during panic:
end of query stack
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'assertion failed: `(left != right)`
  left: `Token { kind: Question, span: default_finding_folder_rust/crash_20210605_162043_0093/delta/perses_result/mutant.rs:2:33: 2:34 (#0) }`,
 right: `Question`', compiler/rustc_parse/src/parser/ty.rs:337:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/dbe459dedd33470f2cb28101157de316caaffa66/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/dbe459dedd33470f2cb28101157de316caaffa66/library/core/src/panicking.rs:92:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
   4: rustc_parse::parser::ty::<impl rustc_parse::parser::Parser>::parse_remaining_bounds
   5: rustc_parse::parser::ty::<impl rustc_parse::parser::Parser>::parse_remaining_bounds_path
   6: rustc_parse::parser::ty::<impl rustc_parse::parser::Parser>::parse_ty_common
   7: rustc_parse::parser::ty::<impl rustc_parse::parser::Parser>::parse_ty_common
   8: rustc_parse::parser::ty::<impl rustc_parse::parser::Parser>::parse_ty_common
   9: rustc_parse::parser::item::<impl rustc_parse::parser::Parser>::parse_param_general::{{closure}}
  10: rustc_parse::parser::item::<impl rustc_parse::parser::Parser>::parse_fn_params::{{closure}}
  11: rustc_parse::parser::item::<impl rustc_parse::parser::Parser>::parse_fn_decl
  12: rustc_parse::parser::item::<impl rustc_parse::parser::Parser>::parse_fn
  13: rustc_parse::parser::item::<impl rustc_parse::parser::Parser>::parse_item_kind
  14: rustc_parse::parser::item::<impl rustc_parse::parser::Parser>::parse_item_common
  15: rustc_parse::parser::item::<impl rustc_parse::parser::Parser>::parse_mod
  16: rustc_parse::parse_crate_from_file
  17: rustc_interface::passes::parse
  18: rustc_interface::queries::Queries::parse
  19: rustc_span::with_session_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.54.0-nightly (dbe459ded 2021-06-02) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z mir-opt-level=3 -Z new-llvm-pass-manager=yes -Z unsound-mir-opts=yes -Z polonius=yes -Z polymorphize=yes -C debuginfo=2 -C opt-level=3 --crate-type staticlib

query stack during panic:
end of query stack
```

</p>
</details>

**Comment by Cancelll** [comment]
seems related to #84148, might be a duplicate.

**Comment by FabianWolff** [comment]
Yes, this seems to have been fixed with #86010 as far as I can tell.

**Comment by JohnTitor** [comment]
Indeed fixed by that PR and it has some regression tests which should also cover this case, so closing. Thanks for reporting!
